File: app/src/test/java/com/jetsetslow/omniterm/RefreshOutcomeTest.kt
package com.jetsetslow.omniterm

import com.jetsetslow.omniterm.data.RefreshHostState
import com.jetsetslow.omniterm.data.RefreshOutcome
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * A pull-to-refresh that ends badly has to say so. The regression this guards is a host left on the
 * "Checking host…" spinner forever: the refresh reported nothing at all, so the app looked broken
 * while the host was merely slow to answer.
 */
class RefreshOutcomeTest {

    private fun host(
        name: String = "atlas",
        status: String = "online",
        probed: Boolean = true,
        authStatus: String = "ok",
        authError: String? = null,
    ) = RefreshHostState(name, status, probed, authStatus, authError)

    @Test
    fun `a healthy fleet reports nothing`() {
        assertNull(RefreshOutcome.describe(listOf(host(), host(name = "beta")), 30))
    }

    @Test
    fun `no hosts reports nothing`() {
        assertNull(RefreshOutcome.describe(emptyList(), 30))
    }

    @Test
    fun `a host still marked connecting is reported, not left silent`() {
        val message = RefreshOutcome.describe(listOf(host(status = "connecting")), 30)
        assertEquals("Refresh problem on 1 host(s): atlas is still not answering after 30s", message)
    }

    @Test
    fun `a host that never completed a probe is reported even when the row says online`() {
        // This is the exact stranding: the status column is not "connecting" but the row still
        // renders the spinner because no probe ever reached a verdict for it.
        assertTrue(RefreshOutcome.isStillChecking(host(status = "online", probed = false)))
        val message = RefreshOutcome.describe(listOf(host(status = "online", probed = false)), 30)
        assertEquals("Refresh problem on 1 host(s): atlas is still not answering after 30s", message)
    }

    @Test
    fun `an offline host is reported as not responding on its route`() {
        assertEquals(
            "Refresh problem on 1 host(s): atlas did not respond on its configured SSH route",
            RefreshOutcome.describe(listOf(host(status = "offline")), 30),
        )
    }

    @Test
    fun `an auth failure surfaces the classified error`() {
        assertEquals(
            "Refresh problem on 1 host(s): atlas: Permission denied (publickey).",
            RefreshOutcome.describe(
                listOf(host(authStatus = "failed", authError = "Permission denied (publickey).")),
                30,
            ),
        )
    }

    @Test
    fun `an auth failure with no message still says something useful`() {
        assertEquals(
            "Refresh problem on 1 host(s): atlas: SSH authentication failed.",
            RefreshOutcome.describe(listOf(host(authStatus = "failed", authError = null)), 30),
        )
    }

    @Test
    fun `still-checking wins over a stale auth failure on the same host`() {
        // Reporting a previous cycle's auth error for a host we could not check this time would be
        // a guess. "Still not answering" is what actually happened.
        assertEquals(
            "Refresh problem on 1 host(s): atlas is still not answering after 30s",
            RefreshOutcome.describe(
                listOf(host(status = "connecting", authStatus = "failed", authError = "old error")),
                30,
            ),
        )
    }

    @Test
    fun `every failing host is named and counted`() {
        val message = RefreshOutcome.describe(
            listOf(
                host(name = "good"),
                host(name = "slow", status = "connecting"),
                host(name = "down", status = "offline"),
            ),
            30,
        )
        assertEquals(
            "Refresh problem on 2 host(s): slow is still not answering after 30s; " +
                "down did not respond on its configured SSH route",
            message,
        )
    }
}

File: flutter_app/test/refresh_outcome_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:omniterm/domain/refresh_outcome.dart';

/// Mirror of Kotlin's `RefreshOutcomeTest`. A pull-to-refresh that ends badly has to say so: the
/// regression this guards is a host left on the "Checking host…" spinner forever while the refresh
/// reported nothing at all.
void main() {
  RefreshHostState host({
    String name = 'atlas',
    String status = 'online',
    bool probed = true,
    String authStatus = 'ok',
    String? authError,
  }) => RefreshHostState(
    name: name,
    status: status,
    probed: probed,
    authStatus: authStatus,
    authError: authError,
  );

  test('a healthy fleet reports nothing', () {
    expect(describeRefreshOutcome([host(), host(name: 'beta')], waitedSeconds: 30), isNull);
  });

  test('no hosts reports nothing', () {
    expect(describeRefreshOutcome(const [], waitedSeconds: 30), isNull);
  });

  test('a host still marked connecting is reported, not left silent', () {
    expect(
      describeRefreshOutcome([host(status: 'connecting')], waitedSeconds: 30),
      'Refresh problem on 1 host(s): atlas is still not answering after 30s',
    );
  });

  test('a host that never completed a probe is reported even when the row says online', () {
    // The exact stranding: the status column is not "connecting" but the row still renders the
    // spinner because no probe ever reached a verdict for it.
    expect(isStillChecking(host(probed: false)), isTrue);
    expect(
      describeRefreshOutcome([host(probed: false)], waitedSeconds: 30),
      'Refresh problem on 1 host(s): atlas is still not answering after 30s',
    );
  });

  test('an offline host is reported as not responding on its route', () {
    expect(
      describeRefreshOutcome([host(status: 'offline')], waitedSeconds: 30),
      'Refresh problem on 1 host(s): atlas did not respond on its configured SSH route',
    );
  });

  test('an auth failure surfaces the classified error', () {
    expect(
      describeRefreshOutcome([
        host(authStatus: 'failed', authError: 'Permission denied (publickey).'),
      ], waitedSeconds: 30),
      'Refresh problem on 1 host(s): atlas: Permission denied (publickey).',
    );
  });

  test('an auth failure with no message still says something useful', () {
    expect(
      describeRefreshOutcome([host(authStatus: 'failed')], waitedSeconds: 30),
      'Refresh problem on 1 host(s): atlas: SSH authentication failed.',
    );
  });

  test('still-checking wins over a stale auth failure on the same host', () {
    expect(
      describeRefreshOutcome([
        host(status: 'connecting', authStatus: 'failed', authError: 'old error'),
      ], waitedSeconds: 30),
      'Refresh problem on 1 host(s): atlas is still not answering after 30s',
    );
  });

  test('every failing host is named and counted', () {
    expect(
      describeRefreshOutcome([
        host(name: 'good'),
        host(name: 'slow', status: 'connecting'),
        host(name: 'down', status: 'offline'),
      ], waitedSeconds: 30),
      'Refresh problem on 2 host(s): slow is still not answering after 30s; '
      'down did not respond on its configured SSH route',
    );
  });
}
